zrml/asset-router/src/pallet_impl/multicurrency.rs

```diff
@@ -116,30 +133,46 @@ impl<T: Config> MultiCurrency<T::AccountId> for Pallet<T> {
         who: &T::AccountId,
         amount: Self::Balance,
     ) -> DispatchResult {
-        if let Ok(currency) = T::CurrencyType::try_from(currency_id) {
-            <T::Currencies as MultiCurrency<T::AccountId>>::withdraw(currency, who, amount)
-        } else if let Ok(asset) = T::MarketAssetType::try_from(currency_id) {
-            // Resulting balance can be ignored as `burn_from` ensures that the
-            // requested amount can be burned.
-            T::MarketAssets::burn_from(asset, who, amount).map(|_| ())
+        if let Ok(asset) = T::MarketAssetType::try_from(currency_id) {
+            // Route "pre new asset system" market assets to `CurrencyType`
+            if T::MarketAssets::asset_exists(asset) {
+                // Resulting balance can be ignored as `burn_from` ensures that the
+                // requested amount can be burned.
+                T::MarketAssets::burn_from(asset, who, amount).map(|_| ())
+            } else if let Ok(currency) = T::CurrencyType::try_from(currency_id) {
+                T::Currencies::withdraw(currency, who, amount)
+            } else {
+                T::MarketAssets::burn_from(asset, who, amount).map(|_| ())
+            }
         } else if let Ok(asset) = T::CampaignAssetType::try_from(currency_id) {
             T::CampaignAssets::burn_from(asset, who, amount).map(|_| ())
         } else if let Ok(asset) = T::CustomAssetType::try_from(currency_id) {
             T::CustomAssets::burn_from(asset, who, amount).map(|_| ())
+        } else if let Ok(currency) = T::CurrencyType::try_from(currency_id) {
+            T::Currencies::withdraw(currency, who, amount)
         } else {
             Err(Error::<T>::UnknownAsset.into())
         }
     }
 
     fn can_slash(currency_id: Self::CurrencyId, who: &T::AccountId, value: Self::Balance) -> bool {
-        if let Ok(currency) = T::CurrencyType::try_from(currency_id) {
-            <T::Currencies as MultiCurrency<T::AccountId>>::can_slash(currency, who, value)
-        } else if let Ok(asset) = T::MarketAssetType::try_from(currency_id) {
-            T::MarketAssets::reducible_balance(asset, who, false) >= value
+        if let Ok(asset) = T::MarketAssetType::try_from(currency_id) {
+            // Route "pre new asset system" market assets to `CurrencyType`
+            if T::MarketAssets::asset_exists(asset) {
+                // Resulting balance can be ignored as `burn_from` ensures that the
+                // requested amount can be burned.
```

Comment:
This comment is incorrect.

it invokes [`do_burn`](https://github.com/paritytech/substrate/blob/bcff60a227d455d95b4712b6cb356ce56b1ff672/frame/assets/src/impl_fungibles.rs#L101-L102)

**Note that `best_effort: false`.**

it invokes [`decrease_balance`](https://github.com/paritytech/substrate/blob/033d4e86cc7eff0066cd376b9375f815761d653c/frame/assets/src/functions.rs#L497)

it invokes [`prep_debit`](https://github.com/paritytech/substrate/blob/033d4e86cc7eff0066cd376b9375f815761d653c/frame/assets/src/functions.rs#L537)

[`prep_debit` docstring](https://github.com/paritytech/substrate/blob/033d4e86cc7eff0066cd376b9375f815761d653c/frame/assets/src/functions.rs#L249-L253) clearly states that `best_effort` determines whether less than `amount` shall be accepted. If `false`, then no. It can also be verified at the [respective `ensure!` call](https://github.com/paritytech/substrate/blob/033d4e86cc7eff0066cd376b9375f815761d653c/frame/assets/src/functions.rs#L263)